jump-core/src/main/java/com/bitdecay/jump/BitWorld.java

```diff
@@ -110,64 +115,116 @@ private void internalStep(final float delta) {
+		staticBodies.stream().forEach(body -> {
+			if (body.active) {
+				updateOccupiedSpaces(body);
+			}
+		});
+		/**
+		 * END OF MOVING EVERYTHING
+		 */
+
+		/**
+		 * BUILD COLLISIONS
+		 */
+		dynamicBodies.stream().filter(body -> body.active).forEach(body -> {
+			buildLevelCollisions(body);
+			expireContact(body);
+			findNewContact(body);
+		});
+		kineticBodies.stream().filter(body -> body.active).forEach(body -> buildKineticCollections(body));
+		/**
+		 * END COLLISIONS
+		 */
+
+
 		resolveAndApplyPendingResolutions();
 
-		bodies.parallelStream().filter(body -> body.active && body.stateWatcher != null).forEach(body -> body.stateWatcher.update(body));
+		dynamicBodies.parallelStream().filter(body -> body.active && body.stateWatcher != null).forEach(body -> body.stateWatcher.update(body));
+	}
+
+	public void updateInput(BitBody body, float delta) {
+		if (body.controller != null) {
+			body.controller.update(delta, body);
+		}
+	}
+
+	public void updateDynamics(BitBody body, float delta) {
+		if (body.gravitational) {
+			body.velocity.add(gravity.scale(delta));
+		}
+	}
+
+	public void updateKinetics(BitBody body) {
+		for (BitBody child : body.children) {
+			/*
+			 * we make sure to move the child just slightly less
+			 * than the parent to guarantee that it still
+			 * collides if nothing else influences it's motion
+			 */
+			BitPoint influence = body.lastAttempt.shrink(MathUtils.FLOAT_PRECISION);
+			child.aabb.translate(influence);
+			// the child did attempt to move this additional amount according to our engine
+			child.lastAttempt.add(influence);
+		}
+		body.children.clear();
+	}
+
+	public void moveBody(BitBody body, float delta) {
+		body.velocity.x = Math.min(Math.abs(body.velocity.x), maxSpeed.x) * (body.velocity.x < 0 ? -1 : 1);
+		body.velocity.y = Math.min(Math.abs(body.velocity.y), maxSpeed.y) * (body.velocity.y < 0 ? -1 : 1);
+		body.lastAttempt = body.velocity.scale(delta);
+		body.aabb.translate(body.lastAttempt);
+	}
+
+	public void resetCollisions(BitBody body) {
+		// all dynamicBodies are assumed to be not grounded unless a collision happens this step.
+		body.grounded = false;
+		// all dynamicBodies assumed to be independent unless a lineage collision happens this step.
+		body.parent = null;
 	}
 
 	private void doAddRemoves() {
-		bodies.removeAll(pendingRemoves);
+		dynamicBodies.removeAll(pendingRemoves);
+		kineticBodies.removeAll(pendingRemoves);
+		staticBodies.removeAll(pendingRemoves);
+		pendingRemoves.stream().forEach(body -> contacts.remove(body));
 		pendingRemoves.clear();
 
-		bodies.addAll(pendingAdds);
+		for (BitBody body : pendingAdds) {
+			if (BodyType.DYNAMIC == body.bodyType) {
+				dynamicBodies.add(body);
+			}
+			if (BodyType.KINETIC == body.bodyType) {
+				kineticBodies.add(body);
+			}
+			if (BodyType.STATIC == body.bodyType) {
+				staticBodies.add(body);
+			}
```

Comment:
True... With the current round of changes, you won't really be able to change the body type on the fly any more. I see no reason we can't just make them their own classes and keep the information localized.